web/packages/teleterm/src/services/config/updateJsonSchema.ts

```diff
@@ -0,0 +1,55 @@
+/**
+ * Copyright 2023 Gravitational, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import path from 'path';
+
+import { z } from 'zod';
+
+import zodToJsonSchema from 'zod-to-json-schema';
+
+import { FileStorage } from 'teleterm/services/fileStorage';
+
+export function updateJsonSchema({
+  configSchema,
+  configFile,
+  configJsonSchemaFile,
+}: {
+  configSchema: z.AnyZodObject;
+  configFile: FileStorage;
+  configJsonSchemaFile: FileStorage;
+}): void {
+  //adds $schema field to the original schema to prevent marking it as a not allowed property
+  const configSchemaWithSchemaField = configSchema.extend({
+    $schema: z.string(),
+  });
+  const jsonSchema = zodToJsonSchema(configSchemaWithSchemaField);
+
+  configJsonSchemaFile.replace(jsonSchema);
+  linkToJsonSchemaIfNeeded(
+    configFile,
+    path.basename(configJsonSchemaFile.getFilePath())
+  );
+}
+
+function linkToJsonSchemaIfNeeded(
```

Comment:
I don't like this function for the following reasons:

* It doesn't help with reducing the abstraction layer of the parent function since `updateJsonSchema` calls both `linkToJsonSchemaIfNeeded` and methods directly on `configJsonSchemaFile`/`configSchema`.
* What is named `configFile` in the parent function gets renamed to `configFileStorage`.

The one advantage of it that I see is that it provides a sort of documentation for the code, but even that could be avoided IMHO.

What do you think about this:

```typescript
export function updateJsonSchema({
  schema,
  configFile,
  schemaFile,
}: {
  schema: z.AnyZodObject;
  configFile: FileStorage;
  schemaFile: FileStorage;
}): void {
  const jsonSchema = zodToJsonSchema(
    // Add $schema field to prevent marking it as a not allowed property.
    schema.extend({ $schema: z.string() })
  );
  const schemaFileName = path.basename(schemaFile.getFilePath());
  const schemaFileNameInConfig = configFile.get('$schema');

  schemaFile.replace(jsonSchema);

  if (schemaFileNameInConfig !== schemaFileName) {
    configFile.put('$schema', schemaFileName);
  }
}
```

The body of this new version has 7 meaningful lines of code and IMHO can be moved to `configService.ts`.

(I wish TypeScript let you destructure and provide types at the same time somehow.)